rpc/codec.go

```diff
@@ -0,0 +1,41 @@
+package rpc
+
+import (
+	"io"
+
+	capnp "zombiezen.com/go/capnproto2"
+)
+
+// TODO(performance):  maintain (global?) sync.Pools for (un)packed encoders & decoders.
+
+// Codec is a factory for encoders and decoders.
+type Codec interface {
+	Encode(*capnp.Message) ([]byte, error)
```

Comment:
The way this interface is written, it locks us into an implementation that does a copy of the message before writing it, whereas in principle there's no need to Marhsal() before writing out the data, and with the current implementation we have the option of optimizing that out. Additionally, it's a bit sad that this type:

```go
type MyCodec struct {
        *Encoder
        *Decoder
}
```

...is not a valid `Codec` by embedding semantics alone.  Making the method singatures for Encode and Decode match those of `Encoder` and `Decoder` would solve both problems.

Reply:
Nice catch.  Agree with all points.